shared/src/test/scala/FetchBatchingTests.scala

```diff
@@ -0,0 +1,145 @@
+
+    override val batchExecution = Parallel
+  }
+
+  def fetchBatchedDataSeq(id: Int): Fetch[Int] = Fetch(BatchedDataSeq(id))
+  def fetchBatchedDataPar(id: Int): Fetch[Int] = Fetch(BatchedDataPar(id))
+
+  "A large fetch to a datasource with a maximum batch size is split and executed in sequence" in {
+    val fetch: Fetch[List[Int]] = Fetch.traverse(List.range(1, 6))(fetchBatchedDataSeq)
+    Fetch.runFetch[Future](fetch).map {
+      case (env, res) =>
+        res shouldEqual List(1, 2, 3, 4, 5)
+        totalFetched(env.rounds) shouldEqual 5
+        totalBatches(env.rounds) shouldEqual 2
+        env.rounds.size shouldEqual 3
+    }
+  }
+
+  "A large fetch to a datasource with a maximum batch size is split and executed in parallel" in {
+    val fetch: Fetch[List[Int]] = Fetch.traverse(List.range(1, 6))(fetchBatchedDataPar)
+    Fetch.runFetch[Future](fetch).map {
+      case (env, res) =>
+        res shouldEqual List(1, 2, 3, 4, 5)
+        totalFetched(env.rounds) shouldEqual 5
+        totalBatches(env.rounds) shouldEqual 2
+        env.rounds.size shouldEqual 1
+    }
+  }
+
+  "Fetches to datasources with a maximum batch size should be split and executed in parallel and sequentially" in {
+    val fetch: Fetch[List[Int]] =
+      Fetch.traverse(List.range(1, 6))(fetchBatchedDataPar) *>
+        Fetch.traverse(List.range(1, 6))(fetchBatchedDataSeq)
+
+    Fetch.runFetch[Future](fetch).map {
+      case (env, res) =>
+        res shouldEqual List(1, 2, 3, 4, 5)
+        totalFetched(env.rounds) shouldEqual 5 + 5
+        totalBatches(env.rounds) shouldEqual 2 + 2
+        env.rounds.size shouldEqual 3
+    }
+  }
+
+  "A large (many) fetch to a datasource with a maximum batch size is split and executed in sequence" in {
+    val fetch: Fetch[List[Int]] =
+      Fetch.multiple(BatchedDataSeq(1), BatchedDataSeq(2), BatchedDataSeq(3))
+    Fetch.runFetch[Future](fetch).map {
+      case (env, res) =>
+        res shouldEqual List(1, 2, 3)
+        totalFetched(env.rounds) shouldEqual 3
+        totalBatches(env.rounds) shouldEqual 2 // FetchMany(NEL(1, 2)) and FetchMany(NEL(3))
+        env.rounds.size shouldEqual 2
+    }
+  }
+
+  "A large (many) fetch to a datasource with a maximum batch size is split and executed in parallel" in {
+    val fetch: Fetch[List[Int]] =
+      Fetch.multiple(BatchedDataPar(1), BatchedDataPar(2), BatchedDataPar(3))
+    Fetch.runFetch[Future](fetch).map {
+      case (env, res) =>
+        res shouldEqual List(1, 2, 3)
+        totalFetched(env.rounds) shouldEqual 3
+        totalBatches(env.rounds) shouldEqual 1
+        env.rounds.size shouldEqual 1
+    }
+  }
+
+  "Very deep fetches don't overflow stack or heap" in {
+    import cats.syntax.traverse._
+
+    val depth = 200
+    val fetch: Fetch[List[Int]] = (1 to depth).toList
+      .map((x) => (0 until 2).toList.traverse(fetchBatchedDataSeq))
+      .foldLeft(
+        Fetch.pure(List.empty[Int])
+      )((acc, f) => acc.flatMap((x) => f))
```

Comment:
This `foldLeft` with `pure` and `flatMap` could also be written using the monadic fold (`Foldable#foldM`).

Something like (might need some type annotations): 
```scala
import cats.instances.list._
import cats.syntax.foldable._

.foldM(List.empty[Int])((_, f) => f)
// or using foldMapM (this does combine all the lists though)
.foldMapM(identity)
```